autode/opt/coordinates/base.py

```diff
@@ -0,0 +1,291 @@
+
+    @abstractmethod
+    def _update_g_from_cart_g(self,
+                              arr: Optional['autode.values.Gradient']
+                              ) -> None:
+        """Update the gradient dE/dR of from a Cartesian (cart) gradient"""
+
+    def update_g_from_cart_g(self,
+                             arr:             Optional['autode.values.Gradient'],
+                             fixed_atom_idxs: Optional[List[int]] = None
+                             ) -> None:
+        """Update the gradient from a Cartesian gradient, zeroing those atoms
+         that are constrained"""
+        assert arr is None or arr.shape[1] == 3  # Needs an Nx3 matrix
+
+        if arr is not None and fixed_atom_idxs is not None:
+            for atom_idx in fixed_atom_idxs:
+                arr[atom_idx, :] = 0.0
+
+        return self._update_g_from_cart_g(arr)
+
+    @abstractmethod
+    def _update_h_from_cart_h(self,
+                              arr: Optional['autode.values.Hessian']
+                              ) -> None:
+        """Update the Hessian from a cartesian Hessian"""
+
+    def update_h_from_cart_h(self,
+                             arr:             Optional['autode.values.Hessian'],
+                             fixed_atom_idxs: Optional[List[int]] = None
+                             ) -> None:
+        """Update the Hessian from a cartesian Hessian with shape 3N x 3N for
+        N atoms, zeroing the second derivatives if required"""
+
+        if arr is not None and fixed_atom_idxs is not None:
+            for atom_idx in fixed_atom_idxs:
+                for i, _ in enumerate(('x', 'y', 'z')):
+                    arr[3*atom_idx+i, :] = arr[:, 3*atom_idx+i] = 0.0
+
+        return self._update_h_from_cart_h(arr)
+
+    def make_hessian_positive_definite(self) -> None:
+        """
+        Make the Hessian matrix positive definite by shifting eigenvalues
+        """
+        hessian = self.h
+
+        if hessian is None:
+            raise RuntimeError('Cannot make a positive definite Hessian - '
+                               'had no Hessian')
+
+        lmd, v = np.linalg.eig(hessian)  # Eigenvalues and eigenvectors
+
+        if np.all(lmd > 0):
+            logger.info('Hessian was positive definite')
+            return
+
+        logger.warning('Hessian was not positive definite. '
+                       'Shifting eigenvalues to 0 and reconstructing')
+        lmd[lmd < 0] = 1E-5
+        self._h = np.linalg.multi_dot((v, np.diag(lmd), v.T)).real
+
+        return None
+
+    @abstractmethod
+    def __repr__(self) -> str:
+        """String representation of these coordinates"""
+
+    @abstractmethod
+    def to(self, *args, **kwargs) -> 'OptCoordinates':
+        """Transformation between these coordinates and another type"""
+
+    @abstractmethod
+    def iadd(self, value: np.ndarray) -> 'OptCoordinates':
+        """Inplace addition of some coordinates"""
+
+    def __eq__(self, other):
```

Comment:
why did you choose to define `__eq__` then?

Reply:
Excellent question... mostly because you get an [LGTM alert](https://lgtm.com/projects/g/duartegroup/autodE/alerts/?mode=list) if `__eq__` isn't overridden in derived classes